### PR TITLE
Redefined POD POC

### PR DESCRIPTION
**Pull request by tanujashinde0405**
# Problem Statement
- Redefined POD POC

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Comment by cortx-admin** [comment]
Can one of the admins verify this patch?

File: test/deploy/kubernetes/provisioner-pods/deployment_data_node1.yaml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+      claimName: dev-sdi-node1
+  containers:
+  - name: cortx-data
+    image: ghcr.io/seagate/cortx-all:2.0.0-latest-custom-ci
+    imagePullPolicy: IfNotPresent
+    args:
+    - /bin/sh
+    - -c
+    - set -x;
+      echo -e "#!/bin/bash\necho $*" > /opt/seagate/cortx/utils/bin/utils_setup;
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/provisioner-pods/deployment_data_node1.yaml:64]: Guess we can control this flow using environment variable to enable/disable mocking...

File: test/deploy/kubernetes/provisioner-pods/deployment_server_node1.yaml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    secret:
+      secretName: cortx-secret
+  - name: shared-node1
+    persistentVolumeClaim:
+      claimName: shared-node1
+  - name: local-node1
+    persistentVolumeClaim:
+      claimName: local-node1
+  - name: sdb
+    persistentVolumeClaim:
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/provisioner-pods/deployment_server_node1.yaml:33]: I doubt if server nodes (pods) need all the devices? Thats only applicable for storage and data nodes (pods)

File: test/deploy/kubernetes/provisioner-pods/deployment_storage_node3.yaml [hunk 3]
```diff
@@ -89,4 +97,4 @@ spec:
     - name: sdh
       devicePath: /dev/sdh
     - name: sdi
-      devicePath: /dev/sdi
+      devicePath: /dev/sdi
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/provisioner-pods/deployment_storage_node3.yaml:100]: Add new line

File: test/deploy/kubernetes/reimage.sh [hunk 4]
```diff
@@ -48,14 +48,6 @@ chmod 700 get_helm.sh;
 print_header "Increasing virtual memory limit";
 sysctl -w vm.max_map_count=30000000;
 
-# Delete previous old images
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/reimage.sh:51]: Why are we removing this logic?

File: test/deploy/kubernetes/runtime-pods/components_data_node2.yaml [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+          mountPath: /etc/cortx
+      - name: cortx-ios1
+        image: ghcr.io/seagate/cortx-all:2.0.0-latest-custom-ci
+        imagePullPolicy: IfNotPresent
+        args:
+        - /bin/sh
+        - -c
+        - set -x;
+          /opt/seagate/cortx/motr/bin/motr_setup start --config yaml:///etc/cortx/cluster.conf --services ioservice --idx 1;
+          sleep infinity;
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/runtime-pods/components_data_node2.yaml:109]: Are we sure to call `sleep infinity` across all pods?

File: test/deploy/kubernetes/runtime-pods/components_server_node1.yaml [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+          sleep infinity;
+        volumeMounts:
+        - name: solution-config
+          mountPath: /etc/cortx/solution
+        - name: node-id
+          mountPath: /etc/cortx/solution/node
+        - name: shared-node1
+          mountPath: /share
+        - name: local-node1
+          mountPath: /etc/cortx
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/runtime-pods/components_server_node1.yaml:183]: Missing New line

File: test/deploy/kubernetes/runtime-pods/components_server_node3.yaml [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+          sleep infinity;
+        volumeMounts:
+        - name: solution-config
+          mountPath: /etc/cortx/solution
+        - name: node-id
+          mountPath: /etc/cortx/solution/node
+        - name: shared-node3
+          mountPath: /share
+        - name: local-node3
+          mountPath: /etc/cortx
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/runtime-pods/components_server_node3.yaml:183]: Missing New line

File: test/deploy/kubernetes/service.sh [hunk 8]
```diff
@@ -3,18 +3,47 @@ BASEPATH=$(dirname $0)
 MAXNODES=$(kubectl get nodes | awk -v col=1 '{print $col}' | tail -n+2 | wc -l)
 NAMESPACE="default"
 TIMEDELAY="5"
+NEW_FLAG=false
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/service.sh:6]: Check above comments

File: test/deploy/kubernetes/upgrade-pods/upgrade_storage_node2.yaml [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    - name: sde
+      devicePath: /dev/sde
+    - name: sdf
+      devicePath: /dev/sdf
+    - name: sdg
+      devicePath: /dev/sdg
+    - name: sdh
+      devicePath: /dev/sdh
+    - name: sdi
+      devicePath: /dev/sdi
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/upgrade-pods/upgrade_storage_node2.yaml:92]: New line

File: test/deploy/kubernetes/upgrade.sh [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+            printf "${FAILED}FAILED${NC}\n"
+            exit 1;
+        else
+            printf "${PASSED}PASSED${NC}\n"
+        fi
+    fi
+done <<< "$(kubectl get pods --namespace=$namespace | grep 'data\|control\|server-node\|ha\|storage')"
+
+# Update Image in Upgrade PODs
+print_header "Updating Upgrade image in Upgrade Pods - Cluster";
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/upgrade.sh:62]: Change message:
`Updating image using Upgrade Pods`

File: test/deploy/kubernetes/destroy.sh [hunk 11]
```diff
@@ -11,11 +11,20 @@ function print_header {
     echo "--------------------------------------------------------------------------"
 }
 
-# Delete Storage Service (Headless)
+# Delete Storage-Node Service (Headless)
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/destroy.sh:14]: Here we should destroy storage or server/data nodes based on infra used...
Here you are only cleaning up data nodes

File: test/deploy/kubernetes/destroy.sh [hunk 12]
```diff
@@ -24,18 +33,52 @@ done
 for NODE_INDEX in $(seq 1 $MAXNODES); do
```
Comment (nitin-seagate) [inline test/deploy/kubernetes/destroy.sh:33]: Change comment message (its confusing)

**Review by nitin-seagate — CHANGES_REQUESTED** [review]
General comment for entire PR:
1. Please update messages and comment appropriately (missed at places causes confusion)
2. Please keep simple workflow across install, destroy, update and service:
if redefined pods:
  ha, control, data, server (setup)
else:
 control, storage

File: test/deploy/kubernetes/deploy.sh [hunk 13]
```diff
@@ -5,13 +5,70 @@ NAMESPACE="default"
...
     echo -e "--------------------------------------------------------------------------"
     echo -e "$1"
     echo -e "--------------------------------------------------------------------------"
 }
 
+function show_usage {
+    echo -e "usage: $(basename $0) [--new] [-i]"
+    echo -e "Where:"
+    echo -e "..."
+    echo -e "IMAGE : CORTX-ALL Image Tag"
```
[the remark this answers is not in the document]
Comment (tanujashinde0405) [reply test/deploy/kubernetes/deploy.sh]: done

File: test/deploy/kubernetes/deploy.sh [hunk 14]
```diff
@@ -5,13 +5,70 @@ NAMESPACE="default"
 TIMEDELAY="5"
 INTRDELAY="2"
 DEF3PLOGS="/tmp/3rdparty_services_install.log"
+NEW_FLAG=false
```
[the remark this answers is not in the document]
Comment (tanujashinde0405) [reply test/deploy/kubernetes/deploy.sh]: done

File: test/deploy/kubernetes/deploy.sh [hunk 15]
```diff
@@ -5,13 +5,70 @@ NAMESPACE="default"
...
+    exit 1
+}
+
+while [ $# -gt 0 ];  do
+    case $1 in
+    -i )
+        shift 1
+        DEPLOY_IMAGE=$1
+        ;;
+    --new )
```
[the remark this answers is not in the document]
Comment (tanujashinde0405) [reply test/deploy/kubernetes/deploy.sh]: done

File: test/deploy/kubernetes/deploy.sh [hunk 16]
```diff
@@ -30,6 +87,12 @@ print_header "Configuring 3rdParty Services";
 ./3rdparty-services/deploy_components.sh > $DEF3PLOGS 2>&1;
 kubectl get service --namespace $NAMESPACE;
 
+if [ $NEW_FLAG = true ]; then
+    cp "$BASEPATH"/template-config/cluster_refined.yaml "$BASEPATH"/solution-config/cluster.yaml;
```
[the remark this answers is not in the document]
Comment (tanujashinde0405) [reply test/deploy/kubernetes/deploy.sh]: done

File: test/deploy/kubernetes/service.sh (hunk 8 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/service.sh]: done

File: test/deploy/kubernetes/upgrade-pods/upgrade_storage_node2.yaml (hunk 9 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/upgrade-pods/upgrade_storage_node2.yaml]: done

File: test/deploy/kubernetes/upgrade.sh (hunk 10 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/upgrade.sh]: done

File: test/deploy/kubernetes/destroy.sh (hunk 12 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/destroy.sh:33]: done

File: test/deploy/kubernetes/runtime-pods/components_server_node3.yaml (hunk 7 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/runtime-pods/components_server_node3.yaml]: done

File: test/deploy/kubernetes/deploy.sh [hunk 17]
```diff
@@ -93,24 +156,79 @@ while [ "$(kubectl get pods | grep control-node | awk '{print $3}')" != "Complet
...
+        print_header "Creating Storage Node - $NODE_NAME";
+        NODE_POD="$BASEPATH/provisioner-pods/deployment_data_$NODE_NAME.yaml";
+        kubectl apply -f "$NODE_POD" --namespace "$NAMESPACE";
+        sleep $TIMEDELAY;
+    done
+
+    # Create Server-Node Service (Headless)
+    for NODE_INDEX in $(seq 1 $MAXNODES); do
+        NODE_NAME="node$NODE_INDEX";
+        print_header "Creating Storage Service - $NODE_NAME";
```
[the remark this answers is not in the document]
Comment (tanujashinde0405) [reply test/deploy/kubernetes/deploy.sh:197]: done

File: test/deploy/kubernetes/deploy.sh [hunk 18]
```diff
@@ -93,24 +156,79 @@ while [ "$(kubectl get pods | grep control-node | awk '{print $3}')" != "Complet
...
+        NODE_SVC="$BASEPATH/external-services/headless_data_$NODE_NAME.yaml";
+        kubectl apply -f "$NODE_SVC" --namespace "$NAMESPACE";
+        sleep $INTRDELAY;
+    done
+
+    # Create Data Node (POD)
+    for NODE_INDEX in $(seq 1 $MAXNODES); do
+        NODE_NAME="node$NODE_INDEX";
+        print_header "Creating Storage Node - $NODE_NAME";
```
[the remark this answers is not in the document]
Comment (tanujashinde0405) [reply test/deploy/kubernetes/deploy.sh:188]: done

File: test/deploy/kubernetes/deploy.sh [hunk 19]
```diff
@@ -93,24 +156,79 @@ while [ "$(kubectl get pods | grep control-node | awk '{print $3}')" != "Complet
...
+        NODE_POD="$BASEPATH/provisioner-pods/deployment_storage_$NODE_NAME.yaml";
+        kubectl apply -f "$NODE_POD" --namespace "$NAMESPACE";
+        sleep $TIMEDELAY;
+    done
+fi
 
-echo -e "Monitor Storage/Control Nodes for Completed status...";
+echo -e "Monitor All Pods for Completed status...";
```
[the remark this answers is not in the document]
Comment (tanujashinde0405) [reply test/deploy/kubernetes/deploy.sh]: done

File: test/deploy/kubernetes/provisioner-pods/deployment_server_node1.yaml (hunk 2 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/provisioner-pods/deployment_server_node1.yaml]: done

File: test/deploy/kubernetes/provisioner-pods/deployment_storage_node3.yaml (hunk 3 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/provisioner-pods/deployment_storage_node3.yaml]: done

File: test/deploy/kubernetes/runtime-pods/components_data_node2.yaml (hunk 5 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/runtime-pods/components_data_node2.yaml:109]: It's there in the services framework also, so following the same. Also As we are mocking the upgrade flow, this throws an error in the upgrade, so as of now we can keep this.

File: test/deploy/kubernetes/runtime-pods/components_server_node1.yaml (hunk 6 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/runtime-pods/components_server_node1.yaml]: done

File: test/deploy/kubernetes/destroy.sh (hunk 11 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/destroy.sh]: Cleaning up all pods as of now, Currently, this script is working for both old as well as redefined pods flow.

File: test/deploy/kubernetes/deploy.sh [hunk 20]
```diff
@@ -93,24 +156,79 @@ while [ "$(kubectl get pods | grep control-node | awk '{print $3}')" != "Complet
...
+        print_header "Creating Storage Service - $NODE_NAME";
+        NODE_SVC="$BASEPATH/external-services/headless_server_$NODE_NAME.yaml";
+        kubectl apply -f "$NODE_SVC" --namespace "$NAMESPACE";
+        sleep $INTRDELAY;
+    done
+
+    # Create Server Node (POD)
+    for NODE_INDEX in $(seq 1 $MAXNODES); do
+        NODE_NAME="node$NODE_INDEX";
+        print_header "Creating Storage Node - $NODE_NAME";
```
[the remark this answers is not in the document]
Comment (tanujashinde0405) [reply test/deploy/kubernetes/deploy.sh]: done

File: test/deploy/kubernetes/3rdparty-services/component-packages/symas_openldap.yaml [hunk 21]
```diff
@@ -16,7 +16,7 @@ spec:
         - "/opt/entrypoint.sh"
       args:
         - "--rootdnpassword"
-        - "seagate2"
+        - "seagate1"
```
Comment (ujjwalpl) [inline test/deploy/kubernetes/3rdparty-services/component-packages/symas_openldap.yaml:19]: Why seagate1 ? Can we say root_dn_password ?

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: test/deploy/kubernetes/deploy.sh [hunk 22]
```diff
@@ -5,13 +5,70 @@ NAMESPACE="default"
...
 DEF3PLOGS="/tmp/3rdparty_services_install.log"
+REDEFPODS=false
 
 function print_header {
     echo -e "--------------------------------------------------------------------------"
     echo -e "$1"
     echo -e "--------------------------------------------------------------------------"
 }
 
+function show_usage {
```
Comment (ujjwalpl) [inline test/deploy/kubernetes/deploy.sh:16]: There should be a similar file called upgrade.sh containing upgrade workflow.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: test/deploy/kubernetes/3rdparty-services/component-packages/symas_openldap.yaml (hunk 21 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/3rdparty-services/component-packages/symas_openldap.yaml:19]: Yes we can change it.

File: test/deploy/kubernetes/deploy.sh (hunk 22 above)
Comment (tanujashinde0405) [reply test/deploy/kubernetes/deploy.sh:16]: Its there in the same directory.

**Pull request merged**